### PR TITLE
Worker function from @taskany/issues for reuse

### LINKED ISSUE #765: Worker function from @taskany/issues for reuse

### PR DESCRIPTION
**Pull request by nyet-ty**
Resolve #765

## PR includes

- [x] Feature

### CODE REVIEW

File: src/utils/worker.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+declare interface JsonArray extends Array<JsonValue> {}
+declare type JsonValue = string | number | boolean | JsonObject | JsonArray | null;
+
+export interface Job {
+    id: string;
+    state: string;
+    priority: number;
+    kind: string;
+    data: JsonValue;
+    delay: number | null;
```
Comment (awinogradov) [inline src/utils/worker.ts:21]: Use optional operator instead. Required null? No way.

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/utils/worker.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    return watchedIds.length;
+};
+
+export const worker = async (
+    getNextJob: (state: jobState, watchedIds: string[]) => Promise<Job>,
+    jobUpdate: (id: string, data: UpdateJobData) => Promise<void>,
+    jobDelete: (id: string) => Promise<void>,
+    resolve: { [key: string]: (args: any) => any },
```
Comment (9teen90nine) [inline src/utils/worker.ts:67]: add a comment to disable warning or change to unknown

File: src/utils/worker.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        );
+
+        if (completedCount + scheduledCount > 300) {
+            onQueeTooLong();
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
+        // eslint-disable-next-line no-console
+        console.error(error.message);
```
Comment (9teen90nine) [inline src/utils/worker.ts:158]: there should be onError callback

**Review by 9teen90nine — CHANGES_REQUESTED** [review]

**Review by awinogradov — CHANGES_REQUESTED** [review]
It would be nice to have some example of usage in JSDoc or md

File: docs/workerExample.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+``` ts
+import { worker, Job } from '@taskany/bricks';
+
+const getNextJob = // function to get first job with needed state update Status to pending, lock before updating
+const jobUpdate = // function to update job
+const jobDelete = // function to delete job
+const resolve = // how to resolve jobs of different kinds
+
+const onRetryLimitExeed = // what to do if retryLimit exeeded
+const onQueeTooLong = // what to do if quee too long
```
Comment (9teen90nine) [inline docs/workerExample.md:10]: queue

File: src/utils/worker.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        );
+
+        if (completedCount + scheduledCount > 300) {
+            onQueeTooLong();
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
+        // eslint-disable-next-line no-console
+        onError(error.message);
```
Comment (9teen90nine) [inline src/utils/worker.ts:160]: better to forward full error, not just the message.

**Review by 9teen90nine — APPROVED** [review]

**Review by awinogradov — APPROVED** [review]

**Pull request merged**